examples/key_based_replication/connector.py

```diff
@@ -0,0 +1,154 @@
+                    "('4', 'John', 'William', 'john@fivetran.com', '2024-02-14T22:59:59Z'), "
+                    "('5', 'Ricky', 'Roma', 'ricky@fivetran.com', '2024-03-16T16:40:29Z')  "
+                    "ON CONFLICT (customer_id) "
+                    "DO UPDATE SET "
+                    "first_name = EXCLUDED.first_name, "
+                    "last_name = EXCLUDED.last_name, "
+                    "email = EXCLUDED.email, "
+                    "updated_at = EXCLUDED.updated_at")
+
+    # Executing insert query.
+    conn.execute(insert_query)
+
+
+# Define the update function, which is a required function, and is called by Fivetran during each sync.
+# See the technical reference documentation for more details on the update function
+# https://fivetran.com/docs/connectors/connector-sdk/technical-reference#update
+# The function takes two parameters:
+# - configuration: dictionary contains any secrets or payloads you configure when deploying the connector
+# - state: a dictionary contains whatever state you have chosen to checkpoint during the prior sync
+# The state dictionary is empty for the first sync or for any full re-sync
+def update(configuration: dict, state: dict):
+    # Connect to your database instance instance.
+    conn = duckdb.connect()
+    # This is not required. This is just for example illustration purposes.
+    setup_source_warehouse(conn)
+    # If the cursor is not present in the state, starting from ('2024-01-01T00:00:00Z') to represent incremental syncs.
+    last_updated_at = state["last_updated_at"] if "last_updated_at" in state else '2024-01-01T00:00:00Z'
+
+    # Fetch records from DB sorted in ascending order.
+    query = (f"SELECT customer_id, first_name, last_name, email, updated_at FROM customers WHERE {replication_key} > "
+             f"'{last_updated_at}' ORDER BY {replication_key}")
+    # This log message will only show while debugging.
+    log.fine(f"fetching records from `customer` table modified after {last_updated_at}")
+    result = conn.execute(query).fetchall()
+
+    # Yield an upsert operation to insert/update the row in the "customers" table.
+    for row in result:
+        yield op.upsert(table="customers",
+                        data={
+                            "customer_id": row[0],  # Customer id.
+                            "first_name": row[1],  # First Name.
+                            "last_name": row[2],  # Last name.
+                            "email": row[3],  # Email id.
+                            "updated_at": dt2str(row[4])  # record updated at.
+                        })
+        # Storing `updated_at` of last fetched record
+        last_updated_at = dt2str(row[4])
+
+    # Update the state to the updated_at of the last record.
+    state["last_updated_at"] = last_updated_at
+    # Save the cursor for the next sync by yielding a checkpoint operation.
+    yield op.checkpoint(state)
```

Comment:
We can recommend that, state can be saved with checkpoint after `N` records in case of large set of data.